lib/src/command/add.dart

```diff
@@ -0,0 +1,405 @@
+/// resolution to find the latest version of the package that is compatible with
+/// the other dependencies in `pubpsec.yaml`, and then enter that as the lower
+/// bound in a ^x.y.z constraint.
+///
+/// Currently supports only adding one dependency at a time.
+class AddCommand extends PubCommand {
+  @override
+  String get name => 'add';
+  @override
+  String get description => 'Add a dependency to pubspec.yaml.';
+  @override
+  String get invocation => 'pub add <package>[:<constraint>] [options]';
+  @override
+  String get docUrl => 'https://dart.dev/tools/pub/cmd/pub-add';
+
+  bool get isDev => argResults['dev'];
+  bool get isDryRun => argResults['dry-run'];
+  String get gitUrl => argResults['git-url'];
+  String get gitPath => argResults['git-path'];
+  String get gitRef => argResults['git-ref'];
+  String get hostUrl => argResults['hosted-url'];
+  String get path => argResults['path'];
+  String get sdk => argResults['sdk'];
+
+  bool get hasGitOptions => gitUrl != null || gitRef != null || gitPath != null;
+  bool get hasHostOptions => hostUrl != null;
+
+  AddCommand() {
+    argParser.addFlag('dev',
+        abbr: 'd',
+        negatable: false,
+        help: 'Adds package to the development dependencies instead.');
+
+    argParser.addOption('git-url', help: 'Git URL of the package');
+    argParser.addOption('git-ref',
+        help: 'Git branch or commit to be retrieved');
+    argParser.addOption('git-path', help: 'Path of git package in repository');
+    argParser.addOption('hosted-url', help: 'URL of package host server');
+    argParser.addOption('path', help: 'Local path');
+    argParser.addOption('sdk', help: 'SDK source for package');
+
+    argParser.addFlag('offline',
+        help: 'Use cached packages instead of accessing the network.');
+
+    argParser.addFlag('dry-run',
+        abbr: 'n',
+        negatable: false,
+        help: "Report what dependencies would change but don't change any.");
+
+    argParser.addFlag('precompile',
+        help: 'Precompile executables in immediate dependencies.');
+  }
+
+  @override
+  Future run() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Must specify a package to be added.');
+    }
+
+    final packageInformation = _parsePackage(argResults.rest.first);
+    final package = packageInformation.first;
+
+    /// Perform version resolution in-memory.
+    final updatedPubSpec =
+        await _addPackageToPubspec(entrypoint.root.pubspec, package);
+
+    SolveResult solveResult;
+
+    try {
+      /// Use [SolveType.UPGRADE] to solve for the highest version of [package]
+      /// in case [package] was already a transitive dependency.
+      solveResult = await resolveVersions(
+        SolveType.UPGRADE,
+        cache,
+        Package.inMemory(updatedPubSpec),
+      );
+    } on GitException {
+      dataError('Unable to resolve package "${package.name}" with the given '
+          'git parameters.');
+    } catch (e) {
```

Comment:
Don't catch all, when we can catch specific errors.. that way if suddenly it starts throwing something new we have bug which causes a stracktrace. Rather than having a silent bug that we are never going to discover.

Reply:
Got it!